### ISSUE TITLE
Definition of object method as a static metaclass function not working

### ISSUE THREAD

**Issue by alexnask** [fasterthanlime/rock#306]
Defining an object method as a static function of the metaclass with this: This as a first parameter does not work, as demonstrated by the following

```
Foo: class {
    bar: static func(this: This) {
        "Hello World!" println()
    }
}
foo := Foo new()
foo bar()
```

This snippet yields the following error: ERROR No such function bar() for `Foo`

**Comment by fasterthanlime** [comment]
I'm pretty sure making this work would slow down the compiler. Any particularly good use case behind that feature ?

**Comment by zhaihj** [comment]
In my opinion, static member should always be accessed through Class but not instance.

Accessing them with instances is confusing:
```ooc
Foo: class{
    val: static Int

    init: func

}

bar := Foo new()
bar val// do something with val

baz := Foo new()
baz val// do some other thing with val
```

Programmer may think that `bar val` and `baz val` are two totally different values, but they are the same.

**Comment by fasterthanlime** [comment]
@zhaihj Yeah, the only place where it should be allowed should be completely unqualified, within a type decl, like so:

```ooc
Foo: class {
  THING_GLOBALLY_DONE := static false

  doThing: func {
    if (THING_GLOBALLY_DONE) return
    "Hey, thing!" println()
    THING_GLOBALLY_DONE = true
  }
}
```

**Comment by shamanas** [comment]
@zhaihj
I agree, however in the case of static funcs that take an object of that class as the first argument, I think it makes sense since a non-static method is(/should be?/debatable?) syntactic sugar for exactly that.

I know D takes this idea to the extreme and allows method syntax to call any function (even those defined outside the class/type) on values which are of the same type as their first argument (Universal Function Call Syntax), although I personally think this is taking it too far.

**Comment by davidhesselbom** [comment]
I'm with @zhaihj, what you're trying to do in the code snippet up top shouldn't be allowed, but I also think the error message you get when doing that should be 
> ERROR Calling static function bar() on Foo instance instead of on Foo class

or something that makes it clear you're doing the wrong thing, and not 

> ERROR No such function bar() for Foo

because that's not true and therefore confusing.

**Comment by fasterthanlime** [comment]
Well

> ERROR No such function bar() for Foo

It should say "instance method", not "function". There is such an instance method on "FooClass", and there is a "static method" on "Foo", but no "instance method" bar() on Foo

Agreed, the error message could be improved.
